File: java/src/main/java/ch4_quicksort/BinarySearchRecursive.java
package ch4_quicksort;

public class BinarySearchRecursive {
    public static void main(String[] args) {
        int[] a = {21, 45, 87, 93};

        System.out.println(binarySearchRecursive(a, 93));
        System.out.println(binarySearchRecursive(a, 16));
    }

    public static int binarySearchRecursive(int[] arr, int target) {
        return binarySearchRecursive(arr, target, 0, arr.length -1);
    }

    private static int binarySearchRecursive(int[] arr, int target, int low, int high) {
        // Check if input list is empty or if target does not exist
        if (arr.length == 0 || low > high) {
            return -1;
        }

        int mid = (low + high) / 2;
        int guess = arr[mid];

        if (guess == target) {
            return mid;
        } else if (guess < target) {
            return binarySearchRecursive(arr, target, mid + 1, high);
        } else {
            return binarySearchRecursive(arr, target, low, high - 1);
        }
    }
}

File: python/04_quicksort/binary_search_recursive.py
def binary_search_recursive(arr, target, low, high):
    if low > high or low == len(arr):
        return None

    mid = (low + high) // 2
    if arr[mid] == target:
        return mid
    elif arr[mid] > target:
        high = mid - 1
    else:
        low = mid + 1
    return binary_search_recursive(arr, target, low, high)


if __name__ == "__main__":
    my_list = [1, 3, 5, 7, 9, 11]
    print(binary_search_recursive(my_list, 0, 0, len(my_list)))
